ADD CapsNet example

Soonhwan-Kwon opened the pull request:
## Description ##
This example is MXNet implementation of [CapsNet](https://arxiv.org/abs/1710.09829):  
Sara Sabour, Nicholas Frosst, Geoffrey E Hinton. Dynamic Routing Between Capsules. NIPS 2017
The current best test error is 0.5%

## Checklist ##
### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] For user-facing API changes, API doc string has been updated. For new C++ functions in header files, their functionalities and arguments are well-documented. 
- [x] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- ADD CapsNet example

Soonhwan-Kwon commented:
We're planning to make this capsule net example as gluon also. Is there anything to do please let me know.

piiswrong commented on example/capsnet/README.md:
```diff
@@ -0,0 +1,32 @@
+Sara Sabour, Nicholas Frosst, Geoffrey E Hinton. Dynamic Routing Between Capsules. NIPS 2017
+- The Current best test error is 0.5%  
+
+Due to the permission issue, this example is maintained in this [repository](https://github.com/samsungsds-rnd/capsnet.mxnet) separately.
```
what does this mean?

Soonhwan-Kwon replied:
If you mean the last line, Pull Request can takes days to accept, meanwhile some follow-up changes can be updated to our temporary repository https://github.com/samsungsds-rnd/capsnet.mxnet.

piiswrong commented:
Does the results match the original paper?

Soonhwan-Kwon commented:
The original paper's result was 0.39% and we expect it can be achieved by adding data augmentation(scale shift) like other implementation of other platforms.

Soonhwan-Kwon commented:
I expect that there are many users of MXNet finding implementation of CapsNet. We're in the middle of implement data augmentation and will make effort to follow up the accuracy soon.

Soonhwan-Kwon commented:
@piiswrong We added data augmentation(shifted up to 2 pixels) as on paper, and fixed structure by blocking gradient on both bias(as constant) and inputs_hat before routing ends. Now it achieves test error =0.44% at epoch 40.

Soonhwan-Kwon commented:
@piiswrong  
We updated test error=0.41% at epoch 62 by parameter tuning, and updated readme and add option for decaying factor to ease parameter tuning.  The test error 0.41% and training time is around 36 sec/epoch, and it shows fastest train time among other implementation(2017-11-21).

| Implementation| test err(%) | ※train time/epoch | GPU  Used|
| :---: | :---: | :---: |:---: |
| MXNet | 0.41 | 36 sec | 2 GTX 1080 |
| tensorflow | 0.49 | ※ 10 min | Unknown(4GB Memory) |
| Keras | 0.34 | 55 sec | 2 GTX 1080 Ti |

※ tensorflow implementation's batch size is not 100 but 128 but MXNet and Keras implementation 's batchsize are 100. 

Sorry for repeating update but would you mind noticing us that it can be merged or not? I'm afraid that there is no response since 2 days ago, and it seems that it is not acceptable as example for MXNet.
